torch_xla_py/xla_multiprocessing.py

```diff
@@ -0,0 +1,124 @@
+from __future__ import division
+from __future__ import print_function
+
+import contextlib
+import os
+import socket
+import sys
+import torch.multiprocessing
+import torch_xla
+import torch_xla_py.xla_env_vars as xenv
+import torch_xla_py.xla_model as xm
+import traceback
+
+
+def _find_free_tcp_port():
+  with contextlib.closing(socket.socket(socket.AF_INET,
+                                        socket.SOCK_STREAM)) as s:
+    s.bind(('', 0))
+    return s.getsockname()[1]
+
+
+def _is_tpu_config():
+  for env in [xenv.TPU_CONFIG, xenv.LOCAL_WORKER]:
+    if os.environ.get(env, None) is not None:
+      return True
+  return False
+
+
+def _parse_tpu_config(config):
+  # XRT_TPU_CONFIG='tpu_worker;0;ismz9:25822'
+  parsed = []
+  for worker in config.split('|'):
+    parts = worker.split(';')
+    if len(parts) != 3:
+      raise ValueError('Bad worker syntax: {}'.format(worker))
+    parsed.append((parts[0], int(parts[1]), parts[2]))
+  return parsed
+
+
+def _get_devices_per_worker():
+  return 8
```

Comment:
this reminds me, did we ever figure out if we can do multiprocessing on a single machine?

Reply:
I change this to use:
```Python
return int(os.environ.get(xenv.TPU_NUM_DEVICES, '8'))
```
As we already have that configurable on the C++ side.

Reply:
Multiprocessing must work for a single machine and for POD/Slice.

Reply:
I meant to ask single machine, per core. So, replace dataparallel threading w/ multiprocessing, essentially.

Reply:
One core is allowed. Or all cores. TPU does not allow to create ad-hoc meshes.
We could enable N cores (N < ALL), with replication disabled.
Inference would be fine, but as soon as you train (and hence CrossReplicaSum) you get hanging.

Reply:
yep, so it's the same as before. I was asking because at some point we were going to look into if we could replace multithreading w/ multiprocessing (as that allows faster data loading apparently).

Reply:
Yes, the replication constraints are not changed as far as TPU mesh goes.